src/fluree/db/util/xhttp.cljc

```diff
@@ -221,48 +220,99 @@
+   :too-big      {:code 1009 :reason "Message too big"}
+   :extension    {:code 1010 :reason "Missing extension"}
+   :unexpected   {:code 1011 :reason "Internal error"}
+   :internal     {:code 1012 :reason "Service restart"}
+   :service      {:code 1013 :reason "Try again later"}
+   :gateway      {:code 1014 :reason "Bad gateway"}
+   :tls          {:code 1015 :reason "TLS handshake"}})
+
+
+
+(defn close-websocket
+  "Closes websocket with optional reason-keyword which
+  will utilize the respective status code and reason string
+  from ws-close-status-codes.
+  Status code info:
+  https://www.rfc-editor.org/rfc/rfc6455.html#section-7.1.5"
+  ([ws]
+   #?(:clj  (ws/close! ws)
+      :cljs (.close ws)))
+  ([ws reason-kw]
+   (let [code   (get-in ws-close-status-codes [reason-kw :code] 1000)
+         reason (get-in ws-close-status-codes [reason-kw :reason] "Normal closure")]
+     #?(:clj  (ws/close! ws code reason)
+        :cljs (.close ws code reason)))))
 
 (defn socket-publish-loop
+  "Sends messages out as they appear on pub-chan.
+  If no message has sent out recently, sends a ping message.
+
+  Does not use ws/ping! as in CLJS web browser support for a
+  true ping is limited/non-existent, so just sends a ping event
+  over the normal message channel which has the same keep-alive effect."
   [ws pub-chan]
   (async/go-loop []
-    (let [val (async/<! pub-chan)]
-      (if (nil? val)
-        (log/info "Web socket pub/producer channel closed.")
-        (let [[msg resp-chan] val]
+    (let [ping-chan (async/timeout 20000)
+          [val ch] (async/alts! [pub-chan ping-chan])]
+      (if (and (= ch pub-chan)
+               (nil? val))
+        (do
+          (log/info "Web socket pub/producer channel closed.")
+          (close-websocket ws :going-away))
+        (let [[msg resp-chan] (if (= ch ping-chan)
+                                [(json/stringify {"action" "ping"}) nil]
+                                val)]
           (try*
-            #?(:clj  (ws/-sendText ^WebSocket ws msg)
+            #?(:clj  (ws/send! ws msg)
                :cljs (.send ws msg))
-            (async/put! resp-chan true)
+            (when resp-chan
+              (async/put! resp-chan true)
+              (async/close! resp-chan))
             (catch* e
                     (log/error e "Error sending websocket message:" msg)
                     (async/put! resp-chan false)))
           (recur))))))
 
 
-(defn close-websocket
-  [ws]
-  #?(:clj  (ha/close-websocket ws)
-     :cljs (.close ws)))
-
-
 (defn try-socket
   [url sub-chan pub-chan resp-chan timeout close-fn]
   #?(:clj
-     (let [client (ha/create-client)
-           ws     (ha/websocket client url
-                                :timeout timeout
-                                :close (fn [_ code reason]
-                                         (log/debug "Websocket closed; code" code
+     (let [ws-config {:connect-timeout timeout
```

Comment:
Definitely not for this pr, but I think stuff like this could work nicely with a framework like donut to manage this config and make it more flexible and pluggable.